File: core/kora-app-annotation-processor/src/main/java/io/koraframework/kora/app/annotation/processor/exception/UnresolvedDependencyException.java
package io.koraframework.kora.app.annotation.processor.exception;

import com.palantir.javapoet.TypeName;
import io.koraframework.annotation.processor.common.CommonClassNames;
import io.koraframework.annotation.processor.common.ProcessingError;
import io.koraframework.annotation.processor.common.ProcessingErrorException;
import io.koraframework.kora.app.annotation.processor.DependencyModuleHintProvider;
import io.koraframework.kora.app.annotation.processor.GraphBuilder;
import io.koraframework.kora.app.annotation.processor.component.DependencyClaim;
import io.koraframework.kora.app.annotation.processor.declaration.ComponentDeclaration;

import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class UnresolvedDependencyException extends ProcessingErrorException {
    private final ComponentDeclaration component;
    private final DependencyClaim dependencyClaim;
    private final Deque<GraphBuilder.ResolutionFrame> stack;

    public UnresolvedDependencyException(TypeElement koraApp,
                                         ComponentDeclaration component,
                                         DependencyClaim dependencyClaim,
                                         Deque<GraphBuilder.ResolutionFrame> stack,
                                         List<DependencyModuleHintProvider.Hint> hints) {

        this(component, dependencyClaim, List.of(getError(koraApp, component, dependencyClaim, stack, hints)), stack);
    }

    private static ProcessingError getError(TypeElement koraApp, ComponentDeclaration component, DependencyClaim dependencyClaim, Deque<GraphBuilder.ResolutionFrame> stack, List<DependencyModuleHintProvider.Hint> hints) {
        var errorSource = dependencyClaim.source() == null ? component.source() : dependencyClaim.source();
        return new ProcessingError(constructErrorMessage(koraApp, component, dependencyClaim, stack, hints), errorSource);
    }


    public UnresolvedDependencyException(ComponentDeclaration component,
                                         DependencyClaim dependencyClaim,
                                         List<ProcessingError> errors,
                                         Deque<GraphBuilder.ResolutionFrame> stack) {
        super(errors);
        this.component = component;
        this.dependencyClaim = dependencyClaim;
        this.stack = stack;
    }

    public ComponentDeclaration getComponent() {
        return component;
    }

    public DependencyClaim getDependencyClaim() {
        return dependencyClaim;
    }

    public Deque<GraphBuilder.ResolutionFrame> getStack() {
        return stack;
    }

    private static String constructErrorMessage(TypeElement koraApp, ComponentDeclaration component, DependencyClaim dependencyClaim, Deque<GraphBuilder.ResolutionFrame> stack, List<DependencyModuleHintProvider.Hint> hints) {
        var msg = new StringBuilder();
        msg.append("No component found for dependency:\n  ");
        msg.append(TypeName.get(dependencyClaim.type()));
        if (dependencyClaim.tag() == null) {
            msg.append(" (no tags)");
        } else {
            msg.append(" with ").append(formatTag(dependencyClaim.tag()));
        }

        var requestedMsg = getRequestedMessage(component);
        msg.append("\n\nRequired at:\n  ").append(requestedMsg);
        var source = dependencyClaim.source();
        if (source instanceof VariableElement variableElement) {
            msg.append("\n  parameter: ").append(variableElement.asType()).append(" ").append(variableElement.getSimpleName());
        }

        var treeMsg = getDependencyTreeSimpleMessage(koraApp, stack, component, dependencyClaim);
        msg.append("\n\n").append(treeMsg);
        if (!hints.isEmpty()) {
            msg.append("\n\nHint:");
            for (var hint : hints) {
                msg.append("\n  - ").append(hint.message().strip().replace("\n", "\n    "));
            }
        }
        msg.append("\n\nFix:");
        msg.append("\n  - Add @").append(CommonClassNames.component.simpleName()).append(" to an implementation of ").append(TypeName.get(dependencyClaim.type())).append('.');
        msg.append("\n  - Add a module method that returns ").append(TypeName.get(dependencyClaim.type())).append('.');
        msg.append("\n  - Include a module that provides ").append(TypeName.get(dependencyClaim.type())).append(" in @KoraApp.");
        return msg.toString();
    }

    private static String formatTag(String tag) {
        return "@Tag(" + tag + ".class)";
    }

    private static String getRequestedMessage(ComponentDeclaration declaration) {
        var element = declaration.source();
        var factoryMethod = (ExecutableElement) null;
        var module = (TypeElement) null;
        do {
            if (element instanceof ExecutableElement) {
                factoryMethod = (ExecutableElement) element;
            } else if (element instanceof TypeElement) {
                module = (TypeElement) element;
                break;
            } else if (element == null) {
                continue;
            }
            element = element.getEnclosingElement();
        } while (element != null);

        if (module != null && factoryMethod != null && factoryMethod.getKind() == ElementKind.CONSTRUCTOR) {
            return "%s.%s".formatted(module.getEnclosingElement(), factoryMethod);
        } else {
            return "%s#%s".formatted(module, factoryMethod);
        }
    }

    private static String getDependencyTreeSimpleMessage(TypeElement koraApp,
                                                         Deque<GraphBuilder.ResolutionFrame> stack,
                                                         ComponentDeclaration declaration,
                                                         DependencyClaim dependencyClaim) {
        var msg = new StringBuilder();
        msg.append("Dependency resolution path:");

        var stackFrames = new ArrayList<GraphBuilder.ResolutionFrame>();
        var i = stack.descendingIterator();
        while (i.hasNext()) {
            var iFrame = i.next();
            if (iFrame instanceof GraphBuilder.ResolutionFrame.Root root) {
                stackFrames.add(root);
                break;
            }
            stackFrames.add(iFrame);
        }

        // reversed order
        var delimiterRoot = "\n  @--- ";
        var delimiter = "\n  ^--- ";
        for (int i1 = stackFrames.size() - 1; i1 >= 0; i1--) {
            var iFrame = stackFrames.get(i1);
            if (iFrame instanceof GraphBuilder.ResolutionFrame.Root root) {
                var rootDeclaration = root.componentDeclaration();
                var rootDeclarationAsStr = rootDeclaration.declarationString();
                var koraAppName = koraApp.getQualifiedName().toString();
                if (rootDeclaration instanceof ComponentDeclaration.FromModuleComponent mc && !rootDeclarationAsStr.contains(koraAppName)) {
                    var moduleTypeName = mc.module().element().getQualifiedName().toString();
                    msg.append(delimiterRoot).append(rootDeclarationAsStr.replace(moduleTypeName, koraAppName));
                    msg.append(delimiter).append(rootDeclarationAsStr);
                } else {
                    msg.append(delimiterRoot).append(rootDeclarationAsStr);
                }
            } else {
                var c = (GraphBuilder.ResolutionFrame.Component) iFrame;
                msg.append(delimiter).append(c.declaration().declarationString());
            }
        }

        msg.append(delimiter).append(declaration.declarationString());

        var errorMissing = " [MISSING]";
        if (dependencyClaim.tag() == null) {
            msg.append(delimiter)
                .append(dependencyClaim.type()).append("   ")
                .append(errorMissing)
                .append("\n");
        } else {
            msg.append(delimiter)
                .append(dependencyClaim.type())
                .append("  @Tag(").append(dependencyClaim.tag()).append(".class)   ")
                .append(errorMissing)
                .append("\n");
        }

        return msg.toString();
    }

}

File: core/kora-app-symbol-processor/src/main/kotlin/io/koraframework/kora/app/ksp/exception/UnresolvedDependencyException.kt
package io.koraframework.kora.app.ksp.exception

import com.google.devtools.ksp.isConstructor
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSDeclaration
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSValueParameter
import com.squareup.kotlinpoet.ksp.toTypeName
import io.koraframework.kora.app.ksp.DependencyModuleHintProvider
import io.koraframework.kora.app.ksp.GraphBuilder
import io.koraframework.kora.app.ksp.component.DependencyClaim
import io.koraframework.kora.app.ksp.declaration.ComponentDeclaration
import io.koraframework.ksp.common.CommonClassNames
import io.koraframework.ksp.common.exception.ProcessingError
import io.koraframework.ksp.common.exception.ProcessingErrorException
import java.util.*
import javax.tools.Diagnostic

class UnresolvedDependencyException(
    val stack: Deque<GraphBuilder.ResolutionFrame>,
    val declaration: ComponentDeclaration,
    val dependencyClaim: DependencyClaim,
    hints: List<DependencyModuleHintProvider.Hint>,
) : ProcessingErrorException(listOf(ProcessingError(getMessage(stack, declaration, dependencyClaim, hints).trimIndent(), dependencyClaim.source ?: declaration.source, Diagnostic.Kind.ERROR))) {


    companion object {
        private fun getMessage(
            stack: Deque<GraphBuilder.ResolutionFrame>,
            declaration: ComponentDeclaration,
            dependencyClaim: DependencyClaim,
            hints: List<DependencyModuleHintProvider.Hint>,
        ): String {
            val msg = StringBuilder()
            msg.append("No component found for dependency:\n  ")
            msg.append(dependencyClaim.type.toTypeName())
            if (dependencyClaim.tag == null) {
                msg.append(" (no tags)")
            } else {
                msg.append(" with @Tag(${dependencyClaim.tag}::class)")
            }
            msg.append("\n\nNote:\n  Kotlin nullable and non-nullable types are different dependency keys.")

            val requestedMsg = getRequestedMessage(declaration)
            msg.append("\n\nRequired at:\n  ").append(requestedMsg)
            val source = dependencyClaim.source
            if (source is KSValueParameter) {
                msg.append("\n  parameter: ")
                    .append(source.type.toTypeName())
                    .append(" ")
                    .append(source.name?.asString() ?: "<unnamed>")
            }

            val treeMsg = getDependencyTreeSimpleMessage(stack, declaration, dependencyClaim)
            msg.append("\n\n").append(treeMsg)
            if (hints.isNotEmpty()) {
                msg.append("\n\nHint:")
                for (hint in hints) {
                    msg.append("\n  - ").append(hint.message().trim().replace("\n", "\n    "))
                }
            }
            msg.append("\n\nFix:")
            msg.append("\n  - Add @${CommonClassNames.component.canonicalName} to an implementation of ${dependencyClaim.type.toTypeName()}.")
            msg.append("\n  - Add a module function that returns ${dependencyClaim.type.toTypeName()}.")
            msg.append("\n  - Include a module that provides ${dependencyClaim.type.toTypeName()} in @KoraApp.")
            return msg.toString()
        }


        private fun getRequestedMessage(declaration: ComponentDeclaration): String {
            var element: KSDeclaration? = declaration.source
            var factoryMethod: KSFunctionDeclaration? = null
            var module: KSDeclaration? = null
            do {
                if (element is KSFunctionDeclaration) {
                    factoryMethod = element
                } else if (element is KSClassDeclaration) {
                    module = element
                    break
                } else if (element == null) {
                    continue
                }
                element = element.parentDeclaration
            } while (element != null)

            return if (module != null && factoryMethod != null && factoryMethod.isConstructor()) {
                "${module.qualifiedName!!.asString()}#${factoryMethod}(${
                    factoryMethod.parameters.joinToString(", ") {
                        it.type.toTypeName().toString()
                    }
                })"
            } else {
                "${module!!.qualifiedName!!.asString()}#${factoryMethod!!.qualifiedName!!.asString()}(${
                    factoryMethod.parameters.joinToString(", ") {
                        it.type.toTypeName().toString()
                    }
                })"
            }
        }

        private fun getDependencyTreeSimpleMessage(
            stack: Deque<GraphBuilder.ResolutionFrame>,
            declaration: ComponentDeclaration,
            dependencyClaim: DependencyClaim,
        ): String {
            val msg = StringBuilder()
            msg.append("Dependency resolution path:")

            val stackFrames = mutableListOf<GraphBuilder.ResolutionFrame>()
            val i = stack.descendingIterator()
            while (i.hasNext()) {
                val iFrame: GraphBuilder.ResolutionFrame = i.next()
                if (iFrame is GraphBuilder.ResolutionFrame.Root) {
                    stackFrames.add(iFrame)
                    break
                }
                stackFrames.add(iFrame)
            }

            // reversed order
            val delimiterRoot = "\n  @--- "
            val delimiterOverriden = "\n  ^~~~ "
            val delimiter = "\n  ^--- "
            for (i1 in stackFrames.indices.reversed()) {
                val iFrame = stackFrames[i1]
                if (iFrame is GraphBuilder.ResolutionFrame.Root) {
                    val rootDeclaration = iFrame.declaration
                    val rootDeclarationAsStr = rootDeclaration.declarationString()
                    if (rootDeclaration is ComponentDeclaration.FromModuleComponent) {
                        val currentModuleTypeName = rootDeclaration.module.element.qualifiedName!!.asString()
                        val originalModuleTypeName = if (rootDeclaration.method.findOverridee()?.parentDeclaration != null) {
                            rootDeclaration.method.findOverridee()!!.parentDeclaration!!.qualifiedName!!.asString()
                        } else {
                            rootDeclaration.module.element.qualifiedName!!.asString()
                        }

                        if (currentModuleTypeName != originalModuleTypeName) {
                            msg.append(delimiterRoot).append(rootDeclarationAsStr.replace(originalModuleTypeName, currentModuleTypeName))
                            msg.append(delimiter).append(rootDeclarationAsStr)
                        } else {
                            msg.append(delimiterRoot).append(rootDeclarationAsStr)
                        }
                    } else {
                        msg.append(delimiterRoot).append(rootDeclarationAsStr)
                    }
                } else {
                    val c = iFrame as GraphBuilder.ResolutionFrame.Component
                    if (c.declaration is ComponentDeclaration.FromModuleComponent && c.declaration.isOverriden()) {
                        msg.append(delimiterOverriden).append(c.declaration.declarationString())
                    } else {
                        msg.append(delimiter).append(c.declaration.declarationString())
                    }
                }
            }

            msg.append(delimiter).append(declaration.declarationString())

            val errorMissing = " [MISSING]"
            if (dependencyClaim.tag == null) {
                msg.append(delimiter)
                    .append(dependencyClaim.type.toTypeName()).append("   ")
                    .append(errorMissing)
            } else {
                msg.append(delimiter)
                    .append(dependencyClaim.type.toTypeName())
                    .append("  @Tag(").append(dependencyClaim.tag).append("::class)   ")
                    .append(errorMissing)
            }

            return msg.toString()
        }
    }

    private fun getRequestedMessage(declaration: ComponentDeclaration): String {
        var element: KSDeclaration? = declaration.source
        var factoryMethod: KSFunctionDeclaration? = null
        var module: KSDeclaration? = null
        do {
            if (element is KSFunctionDeclaration) {
                factoryMethod = element
            } else if (element is KSClassDeclaration) {
                module = element
                break
            } else if (element == null) {
                continue
            }
            element = element.parentDeclaration
        } while (element != null)

        return if (module != null && factoryMethod != null && factoryMethod.isConstructor()) {
            "Dependency requested at: ${module.qualifiedName!!.asString()}#${factoryMethod}(${
                factoryMethod.parameters.joinToString(", ") {
                    it.type.toTypeName().toString()
                }
            })"
        } else {
            "Dependency requested at: ${module!!.qualifiedName!!.asString()}#${factoryMethod!!.qualifiedName!!.asString()}(${
                factoryMethod.parameters.joinToString(", ") {
                    it.type.toTypeName().toString()
                }
            })"
        }
    }

}
